### ISSUE TITLE
[Bug]: bit7z doesn't support UTF-8 for zip filenames under Windows

### ISSUE THREAD

**Issue by ra-dave** [rikyoz/bit7z#171]
### bit7z version

4.0

### 7-zip version

v22.01

### 7-zip shared library used

7z.dll / 7z.so

### Compilers

MSVC

### Architecture

x86_64

### Operating system

Windows

### Operating system versions

Windows 10

### Bug description

On Windows, bit7z errors when the zip filename has UTF-8 characters.  The std::exception's ex.what() is:

No mapping for the Unicode character exists in the target multi-byte code page.

### Steps to reproduce

Just try to create a zip file that has UTF-8 characters in the filename.

_No response_

### Code of Conduct

- [X] By submitting this issue, I agree to follow bit7z's [Code of Conduct](https://github.com/rikyoz/bit7z/blob/master/CODE_OF_CONDUCT.md)

**Comment by rikyoz** [comment]
Hi!
Thanks for reporting it!
Which version of MSVC are you using?
I couldn't reproduce your issue, but I only tested with recent versions of MSVC.

However, while troubleshooting the problem, I found a few lines of code where path encoding wasn't correctly handled.
It might be related to your issue.

I just pushed a commit to the `develop` branch that should fix those lines (d789fa664cefc094cc9b869a30ba6292b6d0961b).

**Comment by ra-dave** [comment]
Appreciate the quick response!  I'm using VS 2022.  

I will test our your change tomorrow and let you know.

**Comment by rikyoz** [comment]
> I'm using VS 2022.
Uhm, this is really strange, as I can't replicate the issue...

> Just tried the new code, still getting the same crash with BitFileCompressor compress().

I just pushed some other minor fixes that should make the encoding handling even more robust.
However, at this point, I don't think the problem is in bit7z...

> BitFileExtractor extract() can't handle a UTF-8 character in the zip filename either. That is the more critical one for me at the moment because I can workaround the BitFileCompressor side if I need to by creating a zip file with a temporary filename and then renaming it outside of bit7z. But trying to do something similar on the extract side is much more problematic (might not have permissions to rename the zip file, etc).

I see. Do you have this issue only with zip files or also with other formats like 7z?

> This is where it is crashing when I try extract():
![](https://user-images.githubusercontent.com/24297299/272443404-f7706533-993f-41d2-adcd-7eb670404066.png)

Bit7z internally uses MSVC's implementation of `std::filesystem`. In this particular case, it uses the function `std::filesystem::u8path` as it expects to construct a `path` from a UTF-8 string.

Now, MSVC's `fs::path` internally uses wide strings, so it needs to convert a UTF-8 string to a wide string.

The error message `No mapping for the Unicode character exists in the target multi-byte code page.` makes me think that the error is that the original string you pass to bit7z is not actually UTF-8 encoded and hence contains some invalid UTF-8 sequences, causing the exception.

One possible way to avoid the throwing of the exception would be to use bit7z's own string conversion functions like this:
```cpp
   return fs::path{ bit7z::widen( str ) };
```
However, if the original string is not UTF-8 encoded, this might not result in a correct path encoding.

**Comment by ra-dave** [comment]
My zip filename seems to be properly UTF-8 encoded (based on the fact that it appears in the debugger's watch window correctly when I use ",s8" modifier.

**Comment by rikyoz** [comment]
> My zip filename seems to be properly UTF-8 encoded (based on the fact that it appears in the debugger's watch window correctly when I use ",s8" modifier.

Uhm ok, strange! I'll need to investigate this further, even though without replicating is a bit difficult to understand what the actual problem is!

> I see that you have some automated tests for testing files in the archive that are UNICODE characters but are there any tests for testing the actual archive filename itself? (for Windows)

Actually, yeah, [here](https://github.com/rikyoz/bit7z/blob/f049382df3341bf01d9a744f846517ec819a704c/tests/src/test_bitarchivereader.cpp#L1115) and [here](https://github.com/rikyoz/bit7z/blob/f049382df3341bf01d9a744f846517ec819a704c/tests/src/test_bitarchivereader.cpp#L1159) the automated tests exactly that: reading an archive with Unicode characters in its name!

**Comment by ra-dave** [comment]
(https://github.com/rikyoz/bit7z/blob/f049382df3341bf01d9a744f846517ec819a704c/tests/src/test_bitarchivereader.cpp#L1115) and [here](https://github.com/rikyoz/bit7z/blob/f049382df3341bf01d9a744f846517ec819a704c/tests/src/test_bitarchivereader.cpp#L1159) the automated tests do exactly that: reading an archive with Unicode characters in its name!

Those tests use BIT7Z_NATIVE_STRING, which is fine for hard-coded strings like that but that can't be used for strings that are not hard-coded.

**Comment by ra-dave** [comment]
Those tests use BIT7Z_NATIVE_STRING, which is fine for hard-coded strings but can't be used for anything else.

I just want to be able to write cross-platform code using UTF-8 encoded std::strings.

We are trying to switch from minizip-ng to bit7z.  minizip just worked without any of this issues.

**Comment by ra-dave** [comment]
> Uhm ok, strange! I'll need to investigate this further, even though without replicating is a bit difficult to understand what the actual problem is!

I will put together a minimal project that replicates the problem.

**Comment by rikyoz** [comment]
> It's easy to replicate. Just use a UTF-8 encoded std::string for the archive filename, in Windows.

Yeah, the problem is that when I try to replicate it, it works!

Just for context, I'm testing everything in a clean installation of Windows 10 in a VM with the default regional settings:

|   |   |
|---|---|
| ![](https://github.com/rikyoz/bit7z/assets/1334291/cc5d85e1-c85d-4f47-869d-3de95eecdcf6) | ![](https://github.com/rikyoz/bit7z/assets/1334291/92556b2b-ea91-4cf0-85a8-c6f1dfd05203) |

The only things installed are 7-zip, Visual Studio 2022, CMake, and Git.
I cloned bit7z in the Documents folder, and created a CMake project in a separate folder called `bit7z-test`:

```cmake
cmake_minimum_required( VERSION 3.8.2 FATAL_ERROR )
project( bit7ztest )

set( CMAKE_CXX_STANDARD 11 )
set( CMAKE_CXX_STANDARD_REQUIRED ON )

add_executable( bit7ztest main.cpp )

set( BIT7Z_AUTO_FORMAT ON )
add_subdirectory( ${PROJECT_SOURCE_DIR}/../bit7z/ ${CMAKE_CURRENT_BINARY_DIR}/bit7z )
target_link_libraries( bit7ztest PRIVATE bit7z )
```

The test files (to be compressed) are in the `test_unicode` folder in the project's folder:
![](https://github.com/rikyoz/bit7z/assets/1334291/d829e214-33f2-4486-b718-41d3ec8cd5e7)

I tested the compression, and it works just fine:
```cpp
#include <iostream>

#include <direct.h>

#include <bit7z/bit7zlibrary.hpp>
#include <bit7z/bitexception.hpp>
#include <bit7z/bitfilecompressor.hpp>

using namespace bit7z;

auto main() -> int {
    std::cout << "Current Code Page: " << GetACP() << std::endl;
    _chdir(R"(C:\Users\Riccardo\Documents\bit7z-test\)");

    try {
        Bit7zLibrary lib{ R"(C:\Program Files\7-Zip\7z.dll)" };
        BitFileCompressor compressor{ lib, BitFormat::Zip };
        compressor.compressDirectory( "test_unicode/", "SottoMenù.zip" );
    } catch (const bit7z::BitException& ex) {
        std::cerr << ex.what() << std::endl;
    }
    return 0;
}
```
![](https://github.com/rikyoz/bit7z/assets/1334291/872dc6f6-0f1f-4cd9-891a-e30f2ca12190)
![](https://github.com/rikyoz/bit7z/assets/1334291/082511d3-dc1a-452a-b6e6-6ef1d1bac924)
![](https://github.com/rikyoz/bit7z/assets/1334291/3cd694db-4302-4f0f-af2a-89ce4c992f91)

I tested extraction too, from the same archive created in the previous test:
```cpp
#include <iostream>

#include <direct.h>

#include <bit7z/bit7zlibrary.hpp>
#include <bit7z/bitexception.hpp>
#include <bit7z/bitfileextractor.hpp>

using namespace bit7z;

auto main() -> int {
    std::cout << "Current Code Page: " << GetACP() << std::endl;
    _chdir(R"(C:\Users\Riccardo\Documents\bit7z-test\)");

    try {
        Bit7zLibrary lib{ R"(C:\Program Files\7-Zip\7z.dll)" };
        BitFileExtractor extractor{ lib, BitFormat::Zip };
        extractor.extract( "SottoMenù.zip", "test_extract/" );
    } catch (const bit7z::BitException& ex) {
        std::cerr << ex.what() << std::endl;
    }
    return 0;
}
```
Again, it worked without issues:
![](https://github.com/rikyoz/bit7z/assets/1334291/76211442-cfa8-4c72-8e64-1195bbf39516)
![](https://github.com/rikyoz/bit7z/assets/1334291/09b7e9d2-e161-40ef-a6b5-6fba4cfca2ee)

All the tests were performed using the `develop` branch of bit7z.

**Comment by rikyoz** [comment]
> It's easy to replicate. Just use a UTF-8 encoded std::string for the archive filename, in Windows.

Yeah, the problem is that when I try to replicate it, it works!

Just for context, I'm testing everything in a clean installation of Windows 10 in a VM with the default regional and language settings:

|   |   |
|---|---|
| ![](https://github.com/rikyoz/bit7z/assets/1334291/cc5d85e1-c85d-4f47-869d-3de95eecdcf6) | ![](https://github.com/rikyoz/bit7z/assets/1334291/92556b2b-ea91-4cf0-85a8-c6f1dfd05203) |

The only things installed are 7-zip, Visual Studio 2022, CMake, and Git.
I cloned bit7z in the Documents folder, and created a CMake project in a separate folder called `bit7z-test`:

![](https://github.com/rikyoz/bit7z/assets/1334291/86a29b5f-22f4-4389-b24a-335d35c1f8b6)

```cmake
cmake_minimum_required( VERSION 3.8.2 FATAL_ERROR )
project( bit7ztest )

set( CMAKE_CXX_STANDARD 11 )
set( CMAKE_CXX_STANDARD_REQUIRED ON )

add_executable( bit7ztest main.cpp )

set( BIT7Z_AUTO_FORMAT ON )
add_subdirectory( ${PROJECT_SOURCE_DIR}/../bit7z/ ${CMAKE_CURRENT_BINARY_DIR}/bit7z )
target_link_libraries( bit7ztest PRIVATE bit7z )
target_compile_options( bit7ztest PRIVATE /utf-8 )
```

The test files (to be compressed) are in the `test_unicode` folder in the project's folder:
![](https://github.com/rikyoz/bit7z/assets/1334291/d829e214-33f2-4486-b718-41d3ec8cd5e7)

I tested the compression, and it works just fine:
```cpp
#include <iostream>

#include <direct.h>

#include <bit7z/bit7zlibrary.hpp>
#include <bit7z/bitexception.hpp>
#include <bit7z/bitfilecompressor.hpp>

using namespace bit7z;

auto main() -> int {
    std::cout << "Current Code Page: " << GetACP() << std::endl;
    _chdir(R"(C:\Users\Riccardo\Documents\bit7z-test\)");

    try {
        Bit7zLibrary lib{ R"(C:\Program Files\7-Zip\7z.dll)" };
        BitFileCompressor compressor{ lib, BitFormat::Zip };
        compressor.compressDirectory( "test_unicode/", u8"SottoMenù.zip" );
    } catch (const bit7z::BitException& ex) {
        std::cerr << ex.what() << std::endl;
    }
    return 0;
}
```
![](https://github.com/rikyoz/bit7z/assets/1334291/872dc6f6-0f1f-4cd9-891a-e30f2ca12190)
![](https://github.com/rikyoz/bit7z/assets/1334291/082511d3-dc1a-452a-b6e6-6ef1d1bac924)
![](https://github.com/rikyoz/bit7z/assets/1334291/3cd694db-4302-4f0f-af2a-89ce4c992f91)

I tested extraction too, from the same archive created in the previous test:
```cpp
#include <iostream>

#include <direct.h>

#include <bit7z/bit7zlibrary.hpp>
#include <bit7z/bitexception.hpp>
#include <bit7z/bitfileextractor.hpp>

using namespace bit7z;

auto main() -> int {
    std::cout << "Current Code Page: " << GetACP() << std::endl;
    _chdir(R"(C:\Users\Riccardo\Documents\bit7z-test\)");

    try {
        Bit7zLibrary lib{ R"(C:\Program Files\7-Zip\7z.dll)" };
        BitFileExtractor extractor{ lib, BitFormat::Zip };
        extractor.extract( u8"SottoMenù.zip", "test_extract/" );
    } catch (const bit7z::BitException& ex) {
        std::cerr << ex.what() << std::endl;
    }
    return 0;
}
```
Again, it worked without issues:
![](https://github.com/rikyoz/bit7z/assets/1334291/76211442-cfa8-4c72-8e64-1195bbf39516)
![](https://github.com/rikyoz/bit7z/assets/1334291/09b7e9d2-e161-40ef-a6b5-6fba4cfca2ee)

All the tests were performed using the `develop` branch of bit7z.

**Comment by rikyoz** [comment]
The only way I found to replicate your issue is by forcing the encoding of the string literal to be Windows-1252:

```cpp
compressor.compressDirectory( "test_unicode/", "SottoMen\xF9.zip" );
```

![](https://github.com/rikyoz/bit7z/assets/1334291/68f67249-e873-4187-b8d4-d8904cb2a88e)

**Comment by rikyoz** [comment]
So, I've released a new maintenance version [v4.0.2](https://github.com/rikyoz/bit7z/releases/tag/v4.0.2) that contains all the fixes to the UTF-8 support I made on the `develop` branch.
I don't know if you tested all these changes, but they could fix your issue.
I'm continuing to try to replicate it, but I haven't been lucky so far.

If the issue persists, could you provide a more detailed stack trace of where the exception happens? At least on bit7z's side.
Also, are you using MSVC's [`/utf-8`](https://learn.microsoft.com/en-us/cpp/build/reference/utf-8-set-source-and-executable-character-sets-to-utf-8?view=msvc-170) option, or the [`/source-charset`](https://learn.microsoft.com/en-us/cpp/build/reference/source-charset-set-source-character-set?view=msvc-170) and [`/execution-charset`](https://learn.microsoft.com/en-us/cpp/build/reference/execution-charset-set-execution-character-set?view=msvc-170) options?
Any details you can provide will be really useful; I really appreciate any help you can provide!

**Comment by rikyoz** [comment]
> I'm still testing but it appears 4.0.2 fixed my issue (4.0.1 definitely still had the issue).
> Thanks!

Great, you're welcome!

> I'll let you know if there are any issues but it's looking good.

Thanks!
